feat(coderd/database/dbpurge): add retention for audit logs

mafredri opened the pull request:
Add configurable retention policy for audit logs. The DeleteOldAuditLogs
query excludes deprecated connection events (connect, disconnect, open,
close) which are handled separately by DeleteOldAuditLogConnectionEvents.

Disabled (0) by default.

Depends on #21021
Updates #20743

---

### PR Stack

|  | PR | Title |
|:---:|:---:|:---|
|  | #21021 | feat(coderd): add retention policy configuration |
|  | #21022 | feat(coderd/database/dbpurge): add retention for connection logs |
| 👉 | #21025 | feat(coderd/database/dbpurge): add retention for audit logs |
|  | #21037 | feat(coderd/database/dbpurge): make API keys retention configurable |
|  | #21038 | docs: add data retention documentation |
|  | #21039 | feat: add retention config for `workspace_agent_logs` |

mafredri committed:
feat(coderd/database/dbpurge): add retention for audit logs

Add configurable retention policy for audit logs. The DeleteOldAuditLogs
query excludes deprecated connection events (connect, disconnect, open,
close) which are handled separately by DeleteOldAuditLogConnectionEvents.

Falls back to global retention if audit logs retention is unset.
Disabled (0) by default.

Depends on #21021
Updates #20743

mafredri committed:
fix: add DeleteOldAuditLogs to dbauthz test suite

mafredri committed:
test(coderd/database/dbpurge): convert audit logs tests to table-driven

mafredri committed:
chore: remove global retention fallback for audit logs

Audit logs retention is now explicit - it's enabled when
--audit-logs-retention is set to a non-zero duration, and
disabled when set to 0. No fallback to global retention.

dannykopping reviewed (approved):
LGTM, except for batch size; I won't need to re-review.

mafredri commented on coderd/database/dbpurge/dbpurge.go:
```diff
@@ -28,6 +28,8 @@ const (
 	// locks that could impact concurrent database operations.
 	connectionLogsBatchSize = 1000
+	// Batch size for audit log deletion.
+	auditLogsBatchSize = 1000
```
Nice, wasn't aware of that issue 👍🏻

mafredri committed:
refactor: optimize DeleteOldAuditLogs to use :execrows

Use :execrows instead of :one to simplify the query by removing the
extra CTE wrapper. This lets PostgreSQL return the row count directly
via RowsAffected() instead of requiring an explicit COUNT(*) scan.

Emyrk reviewed (approved):
:+1: 

Agree 1k is low
